table/block_based/block_based_table_iterator.cc

```diff
@@ -278,34 +308,76 @@ void BlockBasedTableIterator::Prev() {
 }
 
 void BlockBasedTableIterator::InitDataBlock() {
-  BlockHandle data_block_handle = index_iter_->value().handle;
+  BlockHandle data_block_handle;
+  bool is_in_cache = false;
+  if (getenv("Print")) {
+    printf("InitDataBlock\n");
+  }
+
+  if (!block_handles_.empty()) {
+    data_block_handle = block_handles_.front()->index_val_.handle;
+    is_in_cache = block_handles_.front()->is_cache_hit_;
+  } else {
+    data_block_handle = index_iter_->value().handle;
+  }
+
   if (!block_iter_points_to_real_block_ ||
       data_block_handle.offset() != prev_block_offset_ ||
       // if previous attempt of reading the block missed cache, try again
       block_iter_.status().IsIncomplete()) {
     if (block_iter_points_to_real_block_) {
       ResetDataIter();
     }
-    auto* rep = table_->get_rep();
 
     bool is_for_compaction =
         lookup_context_.caller == TableReaderCaller::kCompaction;
-    // Prefetch additional data for range scans (iterators).
-    // Implicit auto readahead:
-    //   Enabled after 2 sequential IOs when ReadOptions.readahead_size == 0.
-    // Explicit user requested readahead:
-    //   Enabled from the very first IO when ReadOptions.readahead_size is set.
-    block_prefetcher_.PrefetchIfNeeded(
-        rep, data_block_handle, read_options_.readahead_size, is_for_compaction,
-        /*no_sequential_checking=*/false, read_options_);
-    Status s;
-    table_->NewDataBlockIterator<DataBlockIter>(
-        read_options_, data_block_handle, &block_iter_, BlockType::kData,
-        /*get_context=*/nullptr, &lookup_context_,
-        block_prefetcher_.prefetch_buffer(),
-        /*for_compaction=*/is_for_compaction, /*async_read=*/false, s,
-        /*use_block_cache_for_lookup=*/true);
+
+    // Initialize Data Block From CacheableEntry.
+    if (is_in_cache) {
+      if (getenv("Print")) {
+        if (block_handles_.front()->cachable_entry_.GetCacheHandle() ==
+            nullptr) {
+          printf("Data cache handle empty\n");
+        } else {
+          printf("Data cache handle not empty\n");
+        }
+      }
+      Status s;
+      block_iter_.Invalidate(Status::OK());
+      table_->NewDataBlockIterator<DataBlockIter>(
+          read_options_, (block_handles_.front()->cachable_entry_).As<Block>(),
+          &block_iter_, s);
+    } else {
+      auto* rep = table_->get_rep();
+
+      std::function<void(size_t, size_t&)> readaheadsize_cb = nullptr;
+      if (readahead_cache_lookup_) {
+        readaheadsize_cb = std::bind(
```

Comment:
Does this need to be done every time, or just once when the iterator is created?

Reply:
This is actually passed only one time when FilePrefetchBuffer is created for the first time in BlockPrefetcher. Remaining time it checks if FilePrefetchBuffer object exists and return.

Reply:
I'm planning to improve this code when refactoring FilePrefetchBuffer.